.circleci/config.yml

```diff
@@ -67,90 +134,105 @@ jobs:
+          name: Deploy fala to production
           command: |
-            source lint-env/bin/activate
-            black --check fala
-  test-web:
+            .circleci/deploy_to_kubernetes production ${ECR_DOCKER_REPO_BASE}-webapp ${ECR_DOCKER_REPO_BASE}-nginx 
+      - deploy:
+          name: Notify Slack channel
+          command: .circleci/notify_slack_channel production-live-0
+
+  live-1_build:
     docker:
-      - image: circleci/python:3.4
+      - image: docker:17.03-git
+    environment:
+      DSD_DOCKER_REGISTRY: "registry.service.dsd.io"
+      DSD_DOCKER_IMAGE: "fala"
     steps:
       - checkout
+      - setup_remote_docker:
+          version: 17.03.0-ce
+          docker_layer_caching: true
       - run:
-          name: Setup Python environment
+          name: Login to the DSD Docker registry
           command: |
-            pip install virtualenv
-            virtualenv env
-            source env/bin/activate
-            pip install pip==18.1
-      - restore_cache:
-          keys:
-            - pip-v1-{{ checksum "requirements/base.txt" }}
+            docker login --username $DOCKER_USERNAME --password $DOCKER_PASSWORD --email "${DOCKER_USERNAME}@digital.justice.gov.uk" $DSD_DOCKER_REGISTRY
       - run:
-          name: Install dependencies
+          name: Login to the ECR Docker registry
           command: |
-            source env/bin/activate
-            pip install --requirement requirements/base.txt
-            pip check
-      - save_cache:
-          key: pip-v1-{{ checksum "requirements/base.txt" }}
-          paths:
-            - "~/.cache/pip"
+            apk add --no-cache --no-progress py2-pip
+            pip install awscli
+            ecr_login="$(aws ecr get-login --region $AWS_DEFAULT_REGION --no-include-email)"
+            ${ecr_login}
       - run:
-          name: Run unit tests
+          name: Build Docker image
           command: |
-            source env/bin/activate
-            python manage.py test
+            docker build --tag application:$CIRCLE_SHA1 \
+              --label build.git.sha=$CIRCLE_SHA1 \
+              --label build.git.branch=$CIRCLE_BRANCH \
+              --label build.url=$CIRCLE_BUILD_URL \
+              .
+      - run:
+          name: Validate Python version
+          command: docker run --rm --tty --interactive application:$CIRCLE_SHA1 python --version | grep "3.4"
+      - run:
+          name: Tag and push Docker images
+          command: |
+            .circleci/tag_and_push_docker_image application:$CIRCLE_SHA1 $DSD_DOCKER_REGISTRY/$DSD_DOCKER_IMAGE
+            .circleci/tag_and_push_docker_image application:$CIRCLE_SHA1 ${ECR_DOCKER_REPO_BASE}
 
-  staging_deploy:
+  live-1_staging_deploy:
     docker:
       - image: ${ECR_ENDPOINT}/cloud-platform/tools:circleci
     steps:
       - checkout
       - run:
           name: Initialise Kubernetes staging context
           command: |
+            unset KUBE_ENV_PRODUCTION_CACERT
```

Comment:
Where do these `KUBE_ENV_PRODUCTION_`* env vars come from

Reply:
These are the ones set on the circle ci site, at the moment on the project settings page, for live-0.
In the new live-1 staging flow, we override the project settings env vars with the ones from the `context`, but the context doesn't have the `_PRODUCTION_` vars in yet, just the `_STAGING_` ones.

The cloud platform tools image finds _all_ keys that start with `KUBE_ENV_`, so it was finding the live-0 production ones and trying to auth on live-1 with them

Once we've added the live-1 `_PRODUCTION_` ones to the context, we won't need to unset these. Having all of these in contexts will mean stuff like this doesn't happen in future, but for now we have to deal with the project settings env vars being fallbacks.

Reply:
It sounds like it'd be worth setting up a `context: laa-fala-live-0` to allow composition instead of exclusion. Is that possible to do?

Reply:
It's possible but for it to have any positive effect on the live-1 deploy, we'd have to move the live-0 secrets from the repo to the context. We would also have to modify the live-0 deploy steps to use the context, which is simple enough. It doesn't seem worthwhile for the timescales involved

To clarify, we only needed to exclude them while the live-1 context didn't have keys to override them. It now does, so the unsetting is redundant (but harmless) now. I think I'll remove it, but leave the keys where they are (and only have a context for live-1).

In general, yes, composition is definitely the way to go.